### PR TITLE
Add default consolidated flag for to_zarr

### PR DESCRIPTION
**Pull request by lsetiawan**
## Overview

This PR adds a default of `consolidated=True` for `to_zarr` but also allows user to modify it, per conversation from #233.

### CODE REVIEW

**Comment by lsetiawan** [comment]
@b-reyes this is ready for review... I hope I covered all of the `to_zarr` appropriately. Thanks.

File: echopype/echodata/zarr_combine.py [hunk 1]
```diff
@@ -600,6 +600,7 @@ def write_to_file(
         zarr_group: str,
         region: Dict[str, slice],
         storage_options: Dict[str, Any] = {},
+        consolidated: bool = True,
```
Comment (b-reyes) [inline echopype/echodata/zarr_combine.py:603]: Please add a docstring for this input variable.

File: echopype/echodata/zarr_combine.py [hunk 2]
```diff
@@ -633,6 +634,7 @@ def write_to_file(
                 compute=True,
                 storage_options=storage_options,
                 synchronizer=zarr.ThreadSynchronizer(),
+                consolidated=consolidated,
```
Comment (b-reyes) [inline echopype/echodata/zarr_combine.py:637]: You should not add consolidated here. The metadata has already been created at this point and you are recomputing it.

File: echopype/echodata/zarr_combine.py [hunk 3]
```diff
@@ -830,10 +854,11 @@ def _write_append_dims(
                     compute=True,
                     storage_options=storage_options,
                     synchronizer=zarr.ThreadSynchronizer(),
+                    consolidated=consolidated,
```
Comment (b-reyes) [inline echopype/echodata/zarr_combine.py:857]: You should not add consolidated here. The metadata has already been created at this point and you are recomputing it.

File: echopype/echodata/zarr_combine.py [hunk 4]
```diff
@@ -1022,6 +1064,7 @@ def combine(
             chunks={},
             synchronizer=zarr.ThreadSynchronizer(),
             storage_options=storage_options,
+            backend_kwargs={"consolidated": consolidated},
```
Comment (b-reyes) [inline echopype/echodata/zarr_combine.py:1067]: Why is this necessary?

**Review by b-reyes — CHANGES_REQUESTED** [review]
@lsetiawan I am done reviewing this PR. From what I can see, It looks like you hit all of the `to_zarr` statements. There are a couple of places where `consolidated` should be removed because the consolidation has already occurred.

With these changes, have you tried to run a large number of files in `combine_echodata`? If not, can you please do that to make sure nothing has changed? 

One thing that I noticed is that in functions such as `to_zarr` in `echodata.py` you provide docstrings for `kwargs`. This seems confusing to me. Is there a numpy doc way of specifying docstring parameters that are specifically for `kwargs`?

File: echopype/echodata/zarr_combine.py (hunk 2 above)
Comment (lsetiawan) [inline echopype/echodata/zarr_combine.py:637]:
```suggestion
```

File: echopype/echodata/zarr_combine.py (hunk 1 above)
Comment (lsetiawan) [inline echopype/echodata/zarr_combine.py:603]:
```suggestion
```

File: echopype/echodata/zarr_combine.py (hunk 2 above)
Comment (lsetiawan) [reply echopype/echodata/zarr_combine.py:637]: Aren't you writing the merged dataset into a new zarr file here?

File: echopype/echodata/zarr_combine.py (hunk 4 above)
Comment (lsetiawan) [reply echopype/echodata/zarr_combine.py:1067]: To be explicit that we are opening a consolidated zarr file.

**Comment by lsetiawan** [comment]
> One thing that I noticed is that in functions such as to_zarr in echodata.py you provide docstrings for kwargs. This seems confusing to me. Is there a numpy doc way of specifying docstring parameters that are specifically for kwargs?

I think this was from the implicity of `to_file` function... we should probably change it so that we do list out all the kwargs. 😛

File: echopype/echodata/zarr_combine.py (hunk 2 above)
Comment (b-reyes) [reply echopype/echodata/zarr_combine.py:637]: Within this function we are just filling in values to a zarr store that has already been consolidated. A quick way to make this determination is to see if `region` is an input of `to_zarr`. If `region` is an input, then the metadata should already be constructed.

File: echopype/echodata/zarr_combine.py (hunk 4 above)
Comment (b-reyes) [reply echopype/echodata/zarr_combine.py:1067]: Where/how would we use/access  this information. I am not really familiar with `backend_kwards`.

**Comment by b-reyes** [comment]
> I think this was from the implicity of to_file function... we should probably change it so that we do list out all the kwargs. 😛

Yeah, I think that is best. If you think this is best too, could you create an issue for it and we can deal with it after this release?

**Comment by lsetiawan** [comment]
> With these changes, have you tried to run a large number of files in combine_echodata? If not, can you please do that to make sure nothing has changed?

I won't have time to fully test this 🙈 Maybe just push this off to the next release since it's not an urgent issue.

**Comment by leewujung** [comment]
@lsetiawan @b-reyes : following up on this -- could you connect and make sure the remaining issues get resolved in this week? I think it would be good to add the required docstrings in this PR to not overcrowd the issue list.

File: echopype/echodata/zarr_combine.py [hunk 5]
```diff
@@ -789,6 +809,7 @@ def _write_append_dims(
         zarr_path: str,
         zarr_group: str,
         storage_options: Dict[str, Any] = {},
+        consolidated: bool = True,
```
Comment (lsetiawan) [inline echopype/echodata/zarr_combine.py:812]:
```suggestion
```

File: echopype/echodata/zarr_combine.py [hunk 6]
```diff
@@ -806,6 +827,9 @@ def _write_append_dims(
         storage_options: dict
             Any additional parameters for the storage
             backend (ignored for local paths)
+        consolidated : bool
+            Flag to consolidate zarr metadata.
+            Defaults to ``True``
```
Comment (lsetiawan) [inline echopype/echodata/zarr_combine.py:832]:
```suggestion
```

File: echopype/echodata/zarr_combine.py (hunk 3 above)
Comment (lsetiawan) [inline echopype/echodata/zarr_combine.py:857]:
```suggestion
```

File: echopype/echodata/zarr_combine.py [hunk 7]
```diff
@@ -1005,23 +1021,41 @@ def combine(
...
+            storage_options=storage_options,
+        )
 
         # change filenames numbering to range(len(eds))
         self._modify_prov_filenames(zarr_path, len_eds=len(eds), storage_options=storage_options)
 
+        if consolidated:
+            # consolidate at the end if consolidated flag is True
+            store = fsspec.get_mapper(zarr_path, **storage_options)
+            zarr.consolidate_metadata(store=store)
```
Comment (lsetiawan) [inline echopype/echodata/zarr_combine.py:1050]: @b-reyes I've modified the code so that `consolidated` flag will consolidate the metadata at the end when the file has been created fully.

1. This will ensure that we don't miss any metadata, say if only parts were consolidated.
2. This should make the process of actually writing faster and it doesn't try to consolidate everytime.

**Comment by lsetiawan** [comment]
@b-reyes This PR is ready for another review. I've put in some tests in there to ensure that everything is consolidating properly. One of the biggest changes is that for your combine function, I put the consolidation at the very end so no metadata is missed, and default all the various private function calls to `to_zarr` to not consolidate.

File: echopype/tests/echodata/utils.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+    tree_dict["Sonar"] = setgrouper.set_sonar()
+    tree_dict["Vendor_specific"] = setgrouper.set_vendor()
+
+    tree = DataTree.from_dict(tree_dict, name="root")
+    echodata = EchoData(
+        source_file=file_chk, xml_path=xml_chk, sonar_model=sonar_model
+    )
+    echodata._set_tree(tree)
+    echodata._load_tree()
+    return echodata
```
Comment (lsetiawan) [inline echopype/tests/echodata/utils.py:103]: Starting work with mock Echodata object referenced in https://github.com/OSOceanAcoustics/echopype/issues/791

Comment (lsetiawan) [reply echopype/tests/echodata/utils.py:103]: FYI @leewujung @emiliom

File: echopype/echodata/zarr_combine.py [hunk 9]
```diff
@@ -627,6 +628,7 @@ def write_to_file(
         """
 
         with Lock(lock_name):
+            # Write to region is not consolidated by default
```
Comment (b-reyes) [inline echopype/echodata/zarr_combine.py:631]: I am not seeing this in the docs anywhere. Perhaps you found this by looking at the actual code? Either way, I think it would be best to be explicit here. Could you please add `consolidated=False` in `ds_in.to_zarr`?

File: echopype/echodata/zarr_combine.py (hunk 7 above)
Comment (b-reyes) [reply echopype/echodata/zarr_combine.py:1053]: Sounds good! Thank you for providing this explanation.

File: echopype/tests/conftest.py [hunk 10]
```diff
@@ -28,6 +28,16 @@ def test_path():
     }
 
 
+@pytest.fixture(scope="session")
```
Comment (b-reyes) [inline echopype/tests/conftest.py:31]: From my understanding, you moved `ek60_test_data` here because it was used within `test_echodata_combine` and `test_zarr_combine`. However, this fixture seems out of place. Also, in the future we may want to have  `ek60_test_data` be different for `test_echodata_combine` and `test_zarr_combine`. Would it be better to just copy this into both `test_echodata_combine` and `test_zarr_combine`?

File: echopype/tests/echodata/test_echodata.py [hunk 11]
```diff
@@ -311,6 +319,49 @@ def test_get_dataset(self, converted_zarr):
         assert result is None
         assert isinstance(ed_result, xr.Dataset)
 
+    @pytest.mark.parametrize("consolidated", [True, False])
+    def test_to_zarr_consolidated(self, mock_echodata, consolidated):
+        """
+        Tests to_zarr method. Currently uses a mock echodata object
+        When consolidated=True, every to_zarr call in ea. group will use the flag
```
Comment (b-reyes) [inline echopype/tests/echodata/test_echodata.py:326]:
```suggestion
        Tests to_zarr consolidation. Currently, this test uses a mock EchoData object that only
        has attributes. The consolidated flag provided will be used in every to_zarr call (which 
        is used to write each EchoData group to zarr_path). 
```

File: echopype/tests/echodata/test_echodata.py [hunk 12]
```diff
@@ -311,6 +319,49 @@ def test_get_dataset(self, converted_zarr):
...
+
+        assert zmeta_path.exists() is check
+
+        if check is True:
+            # Check that every group is in
+            # the zmetadata if consolidated
+            expected_zgroups = [
+                ".zgroup",
+                "Environment/.zgroup",
+                "Platform/.zgroup",
```
Comment (b-reyes) [inline echopype/tests/echodata/test_echodata.py:342]: I noticed that `"Platform/NMEA/.zgroup"` is not here. This should be added. I will also make a comment on the mock EchoData creation.

File: echopype/tests/echodata/test_echodata.py [hunk 13]
```diff
@@ -311,6 +319,49 @@ def test_get_dataset(self, converted_zarr):
...
+                "Provenance/.zgroup",
+                "Sonar/.zgroup",
+                "Sonar/Beam_group1/.zgroup",
+                "Vendor_specific/.zgroup"
+            ]
+            import json
```
Comment (b-reyes) [inline echopype/tests/echodata/test_echodata.py:348]: Is there a reason why this import needs to be here? If not, I would suggest moving it to the top of the file.

File: echopype/tests/echodata/test_echodata.py [hunk 14]
```diff
@@ -311,6 +319,49 @@ def test_get_dataset(self, converted_zarr):
...
+            with open(zmeta_path) as f:
+                meta_json = json.load(f)
+
+            file_groups = [
```
Comment (b-reyes) [inline echopype/tests/echodata/test_echodata.py:352]: Could you add a comment here? Something along the lines of "obtain each .zgroup path in zmeta_path".

File: echopype/tests/echodata/test_zarr_combine.py [hunk 15]
```diff
@@ -300,3 +327,101 @@ def test_append_ds_list_to_zarr(append_ds_list_params):
...
+            echodata_filenames=echodata_filenames,
+            consolidated=consolidated,
+        )
+
+        check = True if consolidated else False
+        zmeta_path = Path(zarr_path) / ".zmetadata"
+
+        assert zmeta_path.exists() is check
+
+        if check is True:
```
Comment (b-reyes) [inline echopype/tests/echodata/test_zarr_combine.py:399]: All of my comments from `test_echodata.py` should apply here. From what I can see this is just a copy paste of items in `test_to_zarr_consolidated` could we make this into a function and put it into say `utils.py`?

File: echopype/tests/echodata/utils.py [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+        return ds
+
+    def set_vendor(self) -> xr.Dataset:
+        # TODO: add mock data
+        ds = xr.Dataset(attrs={"created_by": "Mock test"})
+        return ds
+
+
+def get_mock_echodata(
+    output_path='test.zarr',
```
Comment (b-reyes) [inline echopype/tests/echodata/utils.py:70]: Out of curiosity, what was your vision for `output_path`? Would `get_mock_echodata` return None in this case?

Comment (lsetiawan) [inline echopype/tests/echodata/utils.py:70]:
```suggestion
```

File: echopype/tests/echodata/utils.py [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+    xml_chk=None,
+):
+    # Setup tree dictionary
+    tree_dict = {}
+
+    setgrouper = SetGroupsTest(
+        parser_obj=None,
+        input_file=file_chk,
+        xml_path=xml_chk,
+        output_path=output_path,
```
Comment (lsetiawan) [inline echopype/tests/echodata/utils.py:82]:
```suggestion
        output_path=None,
```

File: echopype/tests/echodata/utils.py (hunk 16 above)
Comment (lsetiawan) [reply echopype/tests/echodata/utils.py:70]: Thanks for catching that!

File: echopype/tests/echodata/test_echodata.py (hunk 13 above)
Comment (lsetiawan) [reply echopype/tests/echodata/test_echodata.py:345]: This is because we only use it for this test.

**Comment by lsetiawan** [comment]
@b-reyes Okay. I've implemented your suggestions. Let me know if I miss anything. Thanks!

File: echopype/tests/echodata/utils.py [hunk 18]
```diff
@@ -0,0 +1,142 @@
...
+    # the zmetadata if consolidated
+    expected_zgroups = [
+        os.path.join(p, '.zgroup') if p != 'Top-level' else '.zgroup'
+        for p in echodata.group_paths
+    ]
+
+    with open(zmeta_path) as f:
+        meta_json = json.load(f)
+
+    file_groups = [
```
Comment (b-reyes) [inline echopype/tests/echodata/utils.py:135]: Could you please put a comment here saying what is happening? I suggested one in the previous review.

File: echopype/tests/echodata/utils.py [hunk 19]
```diff
@@ -0,0 +1,142 @@
...
+    tree = DataTree.from_dict(tree_dict, name="root")
+    echodata = EchoData(
+        source_file=file_chk, xml_path=xml_chk, sonar_model=sonar_model
+    )
+    echodata._set_tree(tree)
+    echodata._load_tree()
+    return echodata
+
+
+def check_consolidated(echodata, zmeta_path):
```
Comment (b-reyes) [inline echopype/tests/echodata/utils.py:119]: Could you put docstrings for the function inputs. I could see one being particularly useful for `zmeta_path`.

**Comment by b-reyes** [comment]
@lsetiawan I am done reviewing. I only have minor remarks in regards to documentation/commenting.

File: echopype/tests/echodata/utils.py (hunk 18 above)
Comment (lsetiawan) [reply echopype/tests/echodata/utils.py:135]: It's in the function docstring

**Review by b-reyes — APPROVED** [review]

**Comment by b-reyes** [comment]
@lsetiawan thank you very much for making those changes. All of my concerns have been addressed. Please go ahead and merge these changes in.

I am slightly worried about the RTD failure, but I believe we can resolve that after a merge, if necessary.

**Comment by lsetiawan** [comment]
> I am slightly worried about the RTD failure, but I believe we can resolve that after a merge, if necessary.

The RTD failure had nothing to do with this PR, no stress 😄 See it all works: https://readthedocs.org/projects/echopype/builds/18907478/
